crates/loader/src/local/mod.rs

```diff
@@ -71,6 +71,17 @@ async fn prepare(
 ) -> Result<Application<CoreComponent>> {
     let info = info(raw.info, &src);
 
+    let id_counts: HashMap<String, u8> =
```

Comment:
Suggestion: pull this out into a `contains_duplicate_ids(...) -> bool` function.  This makes it easier to understand at the point where it's called, and means that if people need to iterate on the implementation then it's guaranteed to be localised to that function.

Reply:
Makes sense, thanks for the suggestion.